File: problems/problems_207/solution.py
import solution
from collections import defaultdict, deque


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.canFinish(*test_input)

    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        # 拓扑排序，我们只能从没有prerequire的课程开始
        graph, q, degree = defaultdict(list), deque([]), [0] * numCourses
        for u, v in prerequisites:
            graph[u].append(v)
            degree[v] += 1
        for i in range(numCourses):
            if degree[i] == 0:
                q.append(i)
        explored = 0
        while q:
            course = q.popleft()
            explored += 1
            for nxt in graph[course]:
                degree[nxt] -= 1
                if degree[nxt] == 0:
                    q.append(nxt)
        return explored == numCourses

File: problems/problems_207/Solution.java
package problems.problems_207;

import com.alibaba.fastjson.JSON;
import java.util.*;
import qubhjava.BaseSolution;


public class Solution extends BaseSolution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] degree = new int[numCourses];
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] req: prerequisites) {
            degree[req[0]]++;
            if (!graph.containsKey(req[1])) {
                graph.put(req[1], new ArrayList<>());
            }
            graph.get(req[1]).add(req[0]);
        }
        Deque<Integer> q = new ArrayDeque<>();
        for (int idx = 0; idx < degree.length; idx++) {
            if (degree[idx] == 0) {
                q.offerLast(idx);
            }
        }
        int explored = 0;
        while (!q.isEmpty()) {
            int cur = q.pollFirst();
            explored++;
            if (graph.containsKey(cur)) {
                for (int other: graph.get(cur)) {
                    degree[other]--;
                    if (degree[other] == 0) {
                        q.offerLast(other);
                    }
                }
            }
        }
        return explored == numCourses;
    }

    @Override
    public Object solve(String[] inputJsonValues) {
        int numCourses = Integer.parseInt(inputJsonValues[0]);
		int[][] prerequisites = jsonArrayToInt2DArray(inputJsonValues[1]);
        return JSON.toJSON(canFinish(numCourses, prerequisites));
    }
}
